<a href="https://colab.research.google.com/github/Alirs2002/fake_and_real_news/blob/main/fake_news.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import tensorflow as tf

In [3]:
import zipfile
zipfile = zipfile.ZipFile("archive_5.zip")

In [4]:
zipfile.extractall()

In [5]:
!ls

archive_5.zip  Fake.csv  sample_data  True.csv


In [11]:
real_news = pd.read_csv("True.csv")
fake_news = pd.read_csv("Fake.csv")

In [12]:
real_news.info(),fake_news.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21417 entries, 0 to 21416
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   title    21417 non-null  object
 1   text     21417 non-null  object
 2   subject  21417 non-null  object
 3   date     21417 non-null  object
dtypes: object(4)
memory usage: 669.4+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23481 entries, 0 to 23480
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   title    23481 non-null  object
 1   text     23481 non-null  object
 2   subject  23481 non-null  object
 3   date     23481 non-null  object
dtypes: object(4)
memory usage: 733.9+ KB


(None, None)

now i think it is better to label the dataset

In [13]:
fake_news["label"] = 0
real_news["label"] = 1

now lets explore the datasert

In [14]:
fake_news.head(4)

title  \
0   Donald Trump Sends Out Embarrassing New Year’...   
1   Drunk Bragging Trump Staffer Started Russian ...   
2   Sheriff David Clarke Becomes An Internet Joke...   
3   Trump Is So Obsessed He Even Has Obama’s Name...   

                                                text subject  \
0  Donald Trump just couldn t wish all Americans ...    News   
1  House Intelligence Committee Chairman Devin Nu...    News   
2  On Friday, it was revealed that former Milwauk...    News   
3  On Christmas day, Donald Trump announced that ...    News   

                date  label  
0  December 31, 2017      0  
1  December 31, 2017      0  
2  December 30, 2017      0  
3  December 29, 2017      0

In [20]:
real_news.iloc[1]["text"]

'WASHINGTON (Reuters) - Transgender people will be allowed for the first time to enlist in the U.S. military starting on Monday as ordered by federal courts, the Pentagon said on Friday, after President Donald Trump’s administration decided not to appeal rulings that blocked his transgender ban. Two federal appeals courts, one in Washington and one in Virginia, last week rejected the administration’s request to put on hold orders by lower court judges requiring the military to begin accepting transgender recruits on Jan. 1. A Justice Department official said the administration will not challenge those rulings. “The Department of Defense has announced that it will be releasing an independent study of these issues in the coming weeks. So rather than litigate this interim appeal before that occurs, the administration has decided to wait for DOD’s study and will continue to defend the president’s lawful authority in District Court in the meantime,” the official said, speaking on condition 

In [23]:
real_news.head(5)

title  \
0  As U.S. budget fight looms, Republicans flip t...   
1  U.S. military to accept transgender recruits o...   
2  Senior U.S. Republican senator: 'Let Mr. Muell...   
3  FBI Russia probe helped by Australian diplomat...   
4  Trump wants Postal Service to charge 'much mor...   

                                                text       subject  \
0  WASHINGTON (Reuters) - The head of a conservat...  politicsNews   
1  WASHINGTON (Reuters) - Transgender people will...  politicsNews   
2  WASHINGTON (Reuters) - The special counsel inv...  politicsNews   
3  WASHINGTON (Reuters) - Trump campaign adviser ...  politicsNews   
4  SEATTLE/WASHINGTON (Reuters) - President Donal...  politicsNews   

                 date  label  
0  December 31, 2017       1  
1  December 29, 2017       1  
2  December 31, 2017       1  
3  December 30, 2017       1  
4  December 29, 2017       1

In [24]:
real_news.subject.value_counts()

subject
politicsNews    11272
worldnews       10145
Name: count, dtype: int64

In [25]:
fake_news.subject.value_counts()

subject
News               9050
politics           6841
left-news          4459
Government News    1570
US_News             783
Middle-east         778
Name: count, dtype: int64

In [27]:
real_news.date.value_counts()

date
December 20, 2017      182
December 6, 2017       166
November 30, 2017      162
November 9, 2017       158
October 13, 2017       155
                      ... 
August 6, 2016           1
August 21, 2016          1
September 3, 2016        1
September 11, 2016       1
January 24, 2016         1
Name: count, Length: 716, dtype: int64

i think, in my opinion, it is not essential to have date and title and the subject in our dataset
so i am going to drop them

In [29]:
real_news.drop(columns=["date","title","subject"],axis = 1,inplace=True)

In [30]:
fake_news.drop(columns=["date","title","subject"],axis = 1,inplace=True)

In [31]:
real_news.head()

text  label
0  WASHINGTON (Reuters) - The head of a conservat...      1
1  WASHINGTON (Reuters) - Transgender people will...      1
2  WASHINGTON (Reuters) - The special counsel inv...      1
3  WASHINGTON (Reuters) - Trump campaign adviser ...      1
4  SEATTLE/WASHINGTON (Reuters) - President Donal...      1

now lets concat the real and fake news

In [37]:
data_set = pd.concat([real_news,fake_news],ignore_index=True)
dataset = data_set.sample(frac=1,ignore_index=True)

In [39]:
dataset.head(),dataset.tail(8)

(                                                text  label
 0  WASHINGTON (Reuters) - Republican lawmakers an...      1
 1  WASHINGTON (Reuters) - After U.S. President Do...      1
 2  WINSTON-SALEM, N.C. (Reuters) - A three-judge ...      1
 3  Donald Trump s buffonery, lies, and unhinged a...      0
 4  Members of the Electoral College will not be r...      0,
                                                     text  label
 44890  Sean Spicer s time as press secretary has been...      0
 44891  21st Century Wire says Four people, including ...      0
 44892  SITTWE, Myanmar (Reuters) - Myanmar s de facto...      1
 44893  Vile GOP Senator Tom Cotton made the mistake o...      0
 44894  Trump s wild lies continue unabated as he cont...      0
 44895  BANGKOK/COX S BAZAR, Bangladesh (Reuters) - Ro...      1
 44896  NEW YORK/WASHINGTON (Reuters) - The U.S. gover...      1
 44897  MADRID (Reuters) - A Spanish court has suspend...      1)

In [41]:
dataset.shape

(44898, 2)

In [42]:
from sklearn.model_selection import train_test_split

train_data , test_data = train_test_split(dataset,test_size=0.1)

In [44]:
test_data.shape,train_data.shape

((4490, 2), (40408, 2))

In [46]:
x_train = train_data.drop(columns=["label"])
x_test = test_data.drop(columns=["label"])

y_train = train_data["label"]
y_test = test_data["label"]

In [47]:
x_train

text
26161  KIEV (Reuters) - Several thousand people march...
23272  Following all the idiotic drama of Donald Trum...
27398  This should come as no surprise to most of us,...
15453  Because according to the left, the brave men w...
36753  WASHINGTON (Reuters) - A proposed tax on $2.6 ...
...                                                  ...
43162  It s not looking as though relations between D...
32595  Hillary was seen passing up a bottle of water ...
24181  BERLIN (Reuters) - German Chancellor Angela Me...
11233  Ted Cruz provided a funny moment during the Re...
9539   WASHINGTON (Reuters) - The leaders of the U.S....

[40408 rows x 1 columns]

In [70]:
x_train.shape

(40408, 1)

In [59]:
len_list = x_train["text"].str.split().str.len()

In [60]:
np.mean(len_list)

405.878019204118

In [61]:
np.max(len_list)

8135

In [62]:
np.percentile(len_list,95)

905.0

So,95 percent of news are shorter than 905 tokens
so we should consider this size for our vectorization and embedding layers

In [73]:

from tensorflow.keras.utils import to_categorical

y_train_encoded = to_categorical(y_train)
y_test_encoded = to_categorical(y_test)

x_train_tensor = tf.convert_to_tensor(x_train)
y_train_tensor = tf.convert_to_tensor(y_train_encoded)

x_test_tensor = tf.convert_to_tensor(x_test)
y_test_tensor = tf.convert_to_tensor(y_test_encoded)


train_data = tf.data.Dataset.from_tensor_slices((x_train_tensor,y_train_tensor))
test_data = tf.data.Dataset.from_tensor_slices((x_test_tensor,y_test_tensor))

now lets create our vectorizer for our model

In [83]:
max_tokens = 150000
vectorizer = tf.keras.layers.TextVectorization(max_tokens,
                                               standardize='lower_and_strip_punctuation',
                                               split='whitespace',
                                               output_mode='int',
                                               output_sequence_length=900)

In [84]:
vectorizer.adapt(x_train)

In [85]:
vectorizer.get_vocabulary()

['',
 '[UNK]',
 'the',
 'to',
 'of',
 'a',
 'and',
 'in',
 'that',
 'on',
 's',
 'for',
 'is',
 'said',
 'he',
 'it',
 'with',
 'was',
 'trump',
 'as',
 'his',
 'by',
 'has',
 'be',
 'have',
 'not',
 'from',
 'this',
 'at',
 'are',
 'who',
 'an',
 'they',
 'us',
 'but',
 'would',
 'we',
 'i',
 'president',
 'about',
 'will',
 'their',
 'had',
 'you',
 't',
 'been',
 'people',
 'were',
 'or',
 'more',
 'which',
 'she',
 'after',
 'her',
 'one',
 'if',
 'all',
 'out',
 'what',
 'its',
 'when',
 'also',
 'state',
 'new',
 'up',
 'there',
 'reuters',
 'no',
 'over',
 'donald',
 'our',
 'house',
 'government',
 'can',
 'clinton',
 'states',
 'than',
 'other',
 'obama',
 'so',
 'him',
 'just',
 'republican',
 'some',
 'could',
 'united',
 'told',
 'into',
 'white',
 'against',
 'like',
 'do',
 'because',
 'campaign',
 'them',
 'last',
 'any',
 'news',
 'two',
 'election',
 'time',
 'now',
 'only',
 'party',
 'how',
 'first',
 'former',
 'washington',
 'even',
 'should',
 'being',
 'year',
 '

In [86]:
len(vectorizer.get_vocabulary())

150000

In [89]:
embedding_1 = tf.keras.layers.Embedding(input_dim=max_tokens,
                                      embeddings_initializer='uniform',
                                      output_dim=128)

now lets createaa out model

In [93]:
input_1 = tf.keras.layers.Input(shape=(1,),dtype="string")
x = vectorizer(input_1)
x = embedding_1(x)
x = tf.keras.layers.Conv1D(128,kernel_size=2,strides=1,padding='valid',activation=tf.keras.activations.relu)(x)
x = tf.keras.layers.Dropout(0.1)(x)

x = tf.keras.layers.LSTM(64,activation='tanh',return_sequences=True)(x)
x = tf.keras.layers.Dense(128,activation=tf.keras.activations.relu)(x)
x = tf.keras.layers.Dropout(0.2)(x)
x = tf.keras.layers.Dense(16,activation=tf.keras.activations.relu)(x)

output_1 = tf.keras.layers.Dense(1,activation=tf.keras.activations.sigmoid)(x)

model_1 = tf.keras.Model(input_1,output_1)

In [94]:
model_1.compile(loss='binary_cross_entropy',
                metrics=['accuracy'],
                optimizer=tf.keras.optimizers.Adam())

In [ ]:
history_1 = model_1.fit(x_train,
                        y_train,
                        epochs=5,
                        verbose=1)